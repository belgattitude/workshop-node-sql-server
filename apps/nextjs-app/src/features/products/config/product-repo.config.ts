import { BrandRepo } from '@/features/products/server/brand.repo';
import { ProductRepo } from '@/features/products/server/product.repo';
import { workshopDatasource } from '@/server/config/ds.kysely-mssql.config';

export const productRepo = new ProductRepo({ ds: workshopDatasource });
export const brandRepo = new BrandRepo({ ds: workshopDatasource });
