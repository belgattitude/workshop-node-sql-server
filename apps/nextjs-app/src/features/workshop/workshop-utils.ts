import { openApi } from 'hono-zod-openapi';
import type { z } from 'zod';

export type WorkshopValidator = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params: z.ZodObject<any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  result: z.ZodArray<any> | z.ZodObject<any>;
};

export type ApiResponse<T> = Promise<{
  data: T;
  meta?: {
    affected?: number;
    sql?: string;
    sqlParams?: Record<string, unknown>;
  };
}>;

export const zodOpenApi = <T extends WorkshopValidator>(validator: T) => {
  return openApi({
    request: {
      query: validator.params as T['params'],
    },
    responses: {
      200: validator.result as T['result'],
    },
  });
};
