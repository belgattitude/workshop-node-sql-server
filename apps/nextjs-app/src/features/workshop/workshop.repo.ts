import type {
  DatasourceResult,
  KyselyDatasource,
} from '@flowblade/source-kysely';
import type { DBKyselySqlServer } from '@workshop/db-sqlserver/kysely-types';
import type { z } from 'zod';

import {
  type WorkshopValidators,
  workshopValidators,
} from '@/features/workshop/workshop.validators';
import type { ApiResponse } from '@/features/workshop/workshop-utils';

type Q1Params = z.infer<WorkshopValidators['query1']['params']>;
type Q1Result = z.infer<WorkshopValidators['query1']['result']>;

export class WorkshopRepo<T = KyselyDatasource<DBKyselySqlServer>> {
  private ds: T;
  public static readonly validators = workshopValidators;

  constructor(params: { ds: T }) {
    this.ds = params.ds;
  }

  /**
   * Query 1: return a list of products
   */
  query1 = async (params: Q1Params): ApiResponse<Q1Result> => {
    const { searchName, limit } = params;
    return {
      data: [
        {
          id: 1,
          brand_id: 1,
          name: 'name',
          reference: 'reference',
          brand_name: 'brand_name',
          barcode_ean13: 'ean13',
        },
      ],
    };
  };

  /**
   * Query 2: return a list of products
   */
  query2 = async (params: Q1Params): ApiResponse<Q1Result> => {
    const { searchName, limit } = params;
    return {
      data: [
        {
          id: 1,
          brand_id: 1,
          name: 'name',
          reference: 'reference',
          brand_name: 'brand_name',
          barcode_ean13: 'ean13',
        },
      ],
    } satisfies DatasourceResult<unknown>;
  };
}
