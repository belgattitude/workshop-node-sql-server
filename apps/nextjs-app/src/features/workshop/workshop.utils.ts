import type {
  QueryResultError,
  QueryResultMeta,
} from '@flowblade/source-kysely';
import { openApi } from 'hono-zod-openapi';
import { z } from 'zod';

export type WorkshopValidator = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params: z.ZodObject<any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  result: z.ZodArray<any> | z.ZodObject<any>;
};

const zMeta = z.object({
  affectedRows: z.number().optional(),
  timeMs: z.number().optional(),
  query: z
    .object({
      sql: z.string().optional(),
      params: z.array(z.unknown()).optional(),
    } satisfies Record<keyof QueryResultMeta['query'], z.Schema>)
    .optional(),
} satisfies Record<keyof QueryResultMeta, z.Schema>);

const zError = z.object({
  message: z.string(),
} satisfies Record<keyof QueryResultError['error'], z.Schema>);

const zQueryResult = <T extends z.Schema>(zSchema: T) =>
  z.object({
    success: z.boolean(),
    data: zSchema.optional(),
    error: zError.optional(),
    meta: zMeta.optional(),
  });

export const zodOpenApi = <T extends WorkshopValidator>(validator: T) => {
  return openApi({
    request: {
      query: validator.params as T['params'],
    },
    responses: {
      200: zQueryResult(validator.result),
    },
  });
};
