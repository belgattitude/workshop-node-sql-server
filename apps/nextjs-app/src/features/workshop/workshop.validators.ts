import { z } from 'zod';

import type { WorkshopValidator } from '@/features/workshop/workshop-utils';

const zProductWithBrand = z.object({
  id: z.number(),
  reference: z.string(),
  name: z.string(),
  barcode_ean13: z.string().nullable(),
  brand_id: z.number().nullable(),
  brand_name: z.string().nullable(),
});

type WorkshopValidatorObject = Record<string, WorkshopValidator>;

export const workshopValidators = {
  query1: {
    params: z.object({
      limit: z.coerce.number().optional().default(100),
      searchName: z.string().optional(),
    }),
    result: z.array(zProductWithBrand),
  },
} as const satisfies WorkshopValidatorObject;

export type WorkshopValidators = typeof workshopValidators;
