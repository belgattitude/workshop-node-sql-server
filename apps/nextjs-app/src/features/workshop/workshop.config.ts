import { WorkshopRepo } from '@/features/workshop/workshop.repo';
import { workshopDatasource } from '@/server/config/ds.kysely-sqlserver.config';

export const workshopRepo = new WorkshopRepo({
  ds: workshopDatasource,
});
