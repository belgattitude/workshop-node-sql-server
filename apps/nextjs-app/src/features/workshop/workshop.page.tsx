'use server';

import type { FC } from 'react';

import { QueryResultDebugger } from '@/components/devtools/QueryResultDebugger';
import { productRepo } from '@/features/products/config/product-repo.config';
import { workshopRepo } from '@/features/workshop/workshop.config';

const workshopExamples = {
  example: () => productRepo.search({ searchName: 'bio', limit: 10 }),
  getStarter: () => workshopRepo.getStarter({ limit: 10 }),
  getBrands: () => workshopRepo.getBrands({ limit: 10 }),
  getProducts: () => workshopRepo.getProducts({ searchName: 'bio', limit: 10 }),
  getAdvanced: () => workshopRepo.getAdvanced({ limit: 100 }),
};

export const WorkshopPage: FC = async () => {
  const result = await workshopExamples.getAdvanced();
  return <QueryResultDebugger result={result} />;
};
