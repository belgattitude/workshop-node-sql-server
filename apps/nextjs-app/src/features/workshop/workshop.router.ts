import { Hono } from 'hono';
import { sql } from 'kysely';

import { workshopRepo } from '@/features/workshop/workshop.config';
import { WorkshopRepo } from '@/features/workshop/workshop.repo';
import { zodOpenApi } from '@/features/workshop/workshop.utils';
import { dbKyselySqlServer } from '@/server/config/db.kysely-sqlserver.config';

const app = new Hono();

const validators = WorkshopRepo.validators;

app.get('/query0', async (c) => {
  const sqlRaw = sql` 
    SELECT TOP 10 b.id, b.name
    FROM [common].[brand] as b
  `;
  const result = await sqlRaw.execute(dbKyselySqlServer);
  return c.json({
    data: result.rows,
  });
});

app.get('/brands', zodOpenApi(validators.getBrands), async (c) => {
  const rows = await workshopRepo.getBrands(c.req.valid('query'));
  return c.json(rows);
});

app.get('/products', zodOpenApi(validators.getProducts), async (c) => {
  const rows = await workshopRepo.query2(c.req.valid('query'));
  return c.json(rows);
});

export { app as workshopRouter };
