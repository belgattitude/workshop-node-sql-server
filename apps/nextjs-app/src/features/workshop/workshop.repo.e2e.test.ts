import { workshopRepo } from '@/features/workshop/workshop.config';

describe('WorkshopRepo', () => {
  it('getStarter should match snapshot', async () => {
    const result = await workshopRepo.getStarter({ limit: 10 });
    expect(result.isOk()).toBe(true);
    const spans = result.meta.getSpans().map((span) => {
      return {
        ...span,
        timeMs: 1,
      };
    });
    expect(spans).toMatchSnapshot('spans');
  });
});
