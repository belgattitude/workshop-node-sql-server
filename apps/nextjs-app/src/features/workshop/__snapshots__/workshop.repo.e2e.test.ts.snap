// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WorkshopRepo > getStarter should match snapshot > spans 1`] = `
[
  {
    "affectedRows": 1934,
    "params": [],
    "sql": " 
       SELECT
           -- limit with top
           b.id, b.name
       FROM [common].[brand] as b
       -- WHERE b.name like 'brand name'
    ",
    "timeMs": 1,
    "type": "sql",
  },
]
`;
